File: fur/c/csim/src/qaoa_fur.c
#include <omp.h>

#include <fur.h>
#include <diagonal.h>

#include <qaoa_fur.h>


void apply_qaoa_furxy_ring(double* sv_real, double* sv_imag, double* const gammas, double* const betas, double* const hc_diag, unsigned int n_qubits, size_t n_states, size_t n_layers, size_t n_trotters)
{
    for (size_t i=0; i<n_layers; i++)
    {
        apply_diagonal(sv_real, sv_imag, -0.5*gammas[i], hc_diag, n_states);
        for (size_t j=0; j<n_trotters; j++)
        {
            furxy_ring(sv_real, sv_imag, 0.5*betas[i]/n_trotters, n_qubits, n_states);
        }
    }
}


void apply_qaoa_furxy_complete(double* sv_real, double* sv_imag, double* const gammas, double* const betas, double* const hc_diag, unsigned int n_qubits, size_t n_states, size_t n_layers, size_t n_trotters)
{
    for (size_t i=0; i<n_layers; i++)
    {
        apply_diagonal(sv_real, sv_imag, -0.5*gammas[i], hc_diag, n_states);
        for (size_t j=0; j<n_trotters; j++)
        {
            furxy_complete(sv_real, sv_imag, 0.5*betas[i]/n_trotters, n_qubits, n_states);
        }
    }
}

File: fur/python/qaoa_fur.py
import typing
import numpy as np
from .fur import furxy_ring, furxy_complete

def apply_qaoa_furxy_ring(sv: np.ndarray, gammas: typing.Sequence[float], betas: typing.Sequence[float], hc_diag: np.ndarray, n_qubits: int, n_trotters: int = 1) -> None:
    for gamma, beta in zip(gammas, betas):
        sv *= np.exp(-0.5j * gamma * hc_diag)
        for _ in range(n_trotters):
            furxy_ring(sv, 0.5 * beta / n_trotters, n_qubits)


def apply_qaoa_furxy_complete(sv: np.ndarray, gammas: typing.Sequence[float], betas: typing.Sequence[float], hc_diag: np.ndarray, n_qubits: int, n_trotters: int = 1) -> None:
    for gamma, beta in zip(gammas, betas):
        sv *= np.exp(-0.5j * gamma * hc_diag)
        for _ in range(n_trotters):
            furxy_complete(sv, 0.5 * beta / n_trotters, n_qubits)
